server/play-server/src/main/scala/sttp/tapir/server/play/PlayServerInterpreter.scala

```diff
@@ -50,46 +52,69 @@ trait PlayServerInterpreter {
+          WebSocket.acceptOrResult { header =>
+            getResponse(header, header.withBody(Source.empty))
           }
+        else
+          playServerOptions.defaultActionBuilder.async(streamParser) { request =>
+            getResponse(header, request).flatMap {
+              case Left(result) => Future.successful(result)
+              case Right(_)     => Future.failed(new Exception("Only WebSocket requests accept flows."))
+            }
+          }
+
+      private def getResponse(
+          header: RequestHeader,
+          request: Request[AkkaStreams.BinaryStream]
+      ): Future[Either[Result, Flow[Message, Message, Any]]] = {
+        implicit val bodyListener: BodyListener[Future, PlayResponseBody] = new PlayBodyListener
+        val serverRequest = new PlayServerRequest(header, request)
+        val interpreter = new ServerInterpreter(
+          new PlayRequestBody(request, playServerOptions),
+          new PlayToResponseBody,
+          playServerOptions.interceptors,
+          playServerOptions.deleteFile
+        )
+
+        interpreter(serverRequest, serverEndpoints).map {
+          case RequestResult.Failure(_) =>
+            Left(Result(header = ResponseHeader(StatusCode.NotFound.code), body = HttpEntity.NoEntity))
+          case RequestResult.Response(response: ServerResponse[PlayResponseBody]) =>
+            val headers: Map[String, String] = response.headers
+              .foldLeft(Map.empty[String, List[String]]) { (a, b) =>
+                if (a.contains(b.name)) a + (b.name -> (a(b.name) :+ b.value)) else a + (b.name -> List(b.value))
+              }
+              .map {
+                // See comment in play.api.mvc.CookieHeaderEncoding
+                case (key, value) if key == HeaderNames.SET_COOKIE => (key, value.mkString(";;"))
+                case (key, value)                                  => (key, value.mkString(", "))
+              }
+              .filterNot(allowToSetExplicitly)
+
+            val status = response.code.code
+            response.body match {
+              case Some(Left(flow))    => Right(flow)
+              case Some(Right(entity)) => Left(Result(ResponseHeader(status, headers), entity))
+              case None =>
+                if (serverRequest.method.is(Method.HEAD) && response.contentLength.isDefined)
+                  Left(
+                    Result(ResponseHeader(status, headers), HttpEntity.Streamed(Source.empty, response.contentLength, response.contentType))
+                  )
+                else Left(Result(ResponseHeader(status, headers), HttpEntity.NoEntity))
+            }
         }
       }
     }
   }
 
+  private def isWebSocket(header: RequestHeader): Boolean =
+    (for {
+      connection <- header.headers.get(sttp.model.HeaderNames.Connection)
+      upgrade <- header.headers.get(sttp.model.HeaderNames.Upgrade)
+    } yield connection.equalsIgnoreCase("Upgrade") && upgrade.equalsIgnoreCase("websocket")).getOrElse(false)
```

Comment:
`Upgrade` will contain the subprotocols which may not be `websocket` so I think we just need to check for the presence of this header.

Reply:
I don't think so? I just checked and the protocol is passed in another header called `sec-websocket-protocol`. 

Looking at the spec: 
> If the response lacks an |Upgrade| header field or the |Upgrade|
       header field contains a value that is not an ASCII case-
       insensitive match for the value "websocket", the client MUST
       _Fail the WebSocket Connection_.

Actually I got this logic from here: https://github.com/softwaremill/tapir/blob/41c6bc7ca656869eb9c7eed7f4f850364968c9bd/core/src/main/scala/sttp/tapir/Tapir.scala#L368

Reply:
Ah you are right I was just looking at the top part of this document: https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism#upgrading_to_a_websocket_connection